<a href="https://colab.research.google.com/github/aitormorais/Ejercicios_fisica/blob/main/Ejercicos_fisica_cuantica.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [63]:
from sympy import symbols, integrate, sin, pi, solve, sqrt,diff,simplify,I
from scipy.integrate import quad
import numpy as np
from scipy import constants as const
import sympy as sp
from scipy.constants import h, c, Rydberg,electron_mass


# Física

---



## Teoria cuántica antigua
---

### **Ejercicio 1: Aplicando la Ley del desplazamiento de Wien, calcúlese las longitudes de onda de mayor energía a la temperatura de:**

---
a) 10^6 K

b) 10^3 K



Definimos la constante de desplazamiento de Wien:

In [ ]:
b = 2.8976e-3

Calculamos longitud de onda para a) 10^6 K

In [ ]:
T1=1e6
lambda1 = b/T1

Calculamos longitud de onda para b) 10^3 K

In [ ]:
T2= 1e3
lambda2 =b/T2

Resultado

In [ ]:
print(f"Longitud de onda para 10^6 K: { lambda1 } m")

Longitud de onda para 10^6 K: 2.8976e-09 m


In [ ]:
print(f"Longitud de onda para 10^6 K: {lambda2} m")

Longitud de onda para 10^6 K: 2.8976e-06 m


### **Ejercicio 2:                                                            Las estrellas enanas blancas tienen una temperatura superficial de unos 10^4 K ¿Cuál es la longitudde onda máxima λ máx correspondiente a esa temperatura según la ley de Wien?**

---

In [ ]:
#sabiendo que λ = b/T
resultado = b/(1e4)
print(f"La longitud de onda correspondiente es: {resultado} m")

La longitud de onda correspondiente es: 2.8976e-07 m


### **Ejercicio 3:  El sol tiene su máxima emisión de energía a una longitud de onda de 4,75·(10^-7) m. Considerando que el espectro de radiación del sol es como el de un cuerpo negro. Calcular la temperatura en la superficie del sol**
---

In [ ]:
res = b/4.75e-7

In [ ]:
print(f"La temperatura de la superfice del sol es {res} K")

La temperatura de la superfice del sol es 6100.21052631579 K


### **Ejercicio 4:Explicar en que consiste “La Catástrofe Ultravioleta”**
---

La "Catástrofe Ultravioleta" se refiere a un problema que surgió a finales del siglo XIX y principios del siglo XX cuando los físicos intentaban explicar la radiación de cuerpo negro, que es la luz emitida por un objeto ideal que absorbe toda la radiación que incide sobre él.

Según la física clásica, la ley de Rayleigh-Jeans predecía que la energía radiada por un cuerpo negro a altas frecuencias (en la región del ultravioleta) debería ser infinitamente grande, lo que obviamente no concordaba con las observaciones experimentales. Se esperaba que la energía radiada por el cuerpo negro aumentara sin límite a medida que la frecuencia de la luz se incrementaba, resultando en una cantidad infinita de energía radiada en el rango ultravioleta, lo cual era absurdo y no se observaba en la naturaleza.

El problema fue resuelto por Max Planck con la introducción de la teoría cuántica, que proponía que la energía se emite en paquetes discretos llamados cuantos. Planck mostró que la energía de estos cuantos era proporcional a la frecuencia de la radiación, lo que limitaba la cantidad de energía emitida a altas frecuencias y resultaba en una predicción que coincidía con los resultados experimentales. Esto marcó el nacimiento de la mecánica cuántica y fue un paso fundamental en la historia de la física.

## El Átomo de Bohr
---

### **Ejercicio 5: Determinar la energía y la longitud de onda λ de la raya espectral que tiene la longitud de onda más larga de la serie de Lyman**
---


La formula de la serie lyman es: $(1/λ) = R ((1/(n_1)
^2) - (1/(n_2)^2))$
Donde:

*   R es la constante de Rydberg
*   n_1 y n_2  son los niveles de energia inicial y final



Definimos las constantes

In [ ]:
R = Rydberg  # m^-1, Constante de Rydberg
h = h  # J s, Constante de Planck
c = c  # m/s, Velocidad de la luz

Calculamos la longitud de onda para n = 2

Cálculo longitud de onda

In [ ]:
n = 2
lambda_inverse = R * (1/1**2 - 1/n**2)
lambda_lyman = 1/lambda_inverse

Cálculo de la energía

In [ ]:
E = h * c * lambda_inverse

Resultados

In [ ]:
print(f"Longitud de onda de la raya espectral más larga de la serie de Lyman: {lambda_lyman} m")
print(f"Energía de la raya espectral más larga de la serie de Lyman: {E} J")

Longitud de onda de la raya espectral más larga de la serie de Lyman: 1.2150227341098497e-07 m
Energía de la raya espectral más larga de la serie de Lyman: 1.6349042708276885e-18 J


### **Ejercicio 6 : Haciendo uso de la fórmula de Rydberg obtener las 4 líneas menos energéticas del espectro del átomo de hidrógeno**
---

Función para calcular la longitud de onda utilizando la fórmula de Rydberg

In [ ]:
def calcular_longitud_de_onda(n1, n2):
    return 1 / (R * (1/n1**2 - 1/n2**2)) * 1e9 #convertir a nm

In [ ]:
# Cálculo de las longitudes de onda para n1=2 y n2 variando de 3 a 6
longitudes_de_onda = [calcular_longitud_de_onda(2, n2) for n2 in range(3, 7)]

In [ ]:
for i in longitudes_de_onda: print(f"{i} nm")


656.112276419319 nm
486.0090936439399 nm
433.93669075351784 nm
410.0701727620744 nm


### **Ejercicio 7:**
**Utilizar el átomo de Borh para explicar el principio de correspondencia**

---


El átomo de Bohr, un modelo del átomo propuesto por Niels Bohr, es una representación temprana de la mecánica cuántica y puede usarse para ilustrar el principio de correspondencia. En este modelo, los electrones orbitan el núcleo en órbitas circulares discretas sin emitir radiación, a diferencia de las predicciones de la física clásica, que dirían que un electrón en una órbita circular debería irradiar energía continuamente y espiral hacia el núcleo.

En el átomo de Bohr, las transiciones entre órbitas estacionarias ocurren con la emisión o absorción de fotones con energías específicas. Estas energías están cuantizadas y corresponden a la diferencia de energía entre los niveles de órbita.

El principio de correspondencia se manifiesta de la siguiente manera:

1. **Transiciones de Baja Energía (Pequeños Números Cuánticos):** Para transiciones entre niveles con números cuánticos bajos (es decir, cercanos al núcleo), la energía de los fotones emitidos es significativamente cuantizada, lo cual es un comportamiento puramente cuántico.

2. **Transiciones de Alta Energía (Grandes Números Cuánticos):** Para transiciones entre niveles con números cuánticos muy altos (donde la diferencia de energía entre los niveles consecutivos es muy pequeña), la energía de los fotones emitidos se acerca a un espectro continuo, que es lo que se esperaría según la física clásica.


### **Ejercicio 8: Describir los conceptos de cuantización que aparecen en los postulados de Bohr para explicar el átomo de hidrógeno**
---

Los postulados de Bohr introdujeron conceptos de cuantización para explicar la estructura del átomo de hidrógeno, que eran revolucionarios en el momento de su propuesta en 1913. Estos conceptos incluyen:

* Órbitas Cuantizadas: Bohr postuló que los electrones solo pueden ocupar ciertas órbitas estacionarias alrededor del núcleo sin radiar energía. Estas órbitas están cuantizadas, es decir, solo ciertos valores de radio son permitidos, y cada uno corresponde a un nivel de energía fijo.

* Cuantización del Momento Angular: Según Bohr, el momento angular de un electrón en una órbita permitida está cuantizado y es un múltiplo entero de $ \hbar/2\pi $, donde $ \hbar $ es la constante de Planck reducida. Es decir, $ L = n\hbar $ donde $ n $ es el número cuántico principal.

* Transiciones y Foton Emisión/Absorción: Bohr también postuló que un electrón puede hacer una transición entre dos órbitas estacionarias. Al hacerlo, la diferencia de energía entre las órbitas se emite o se absorbe en forma de un fotón con una energía $ E = h\nu $, donde $ h $ es la constante de Planck y $ \nu $ es la frecuencia del fotón. Esto significa que la energía radiada o absorbida está también cuantizada.

* Frecuencias de Espectros: El modelo explicaba las líneas espectrales observadas para el átomo de hidrógeno. Las frecuencias de las líneas espectrales están determinadas por la diferencia de energía entre las órbitas cuantizadas, lo que concuerda con la fórmula de Rydberg y la serie de Balmer observada experimentalmente para el hidrógeno.






## Dualidad Onda-Corpúsculo
---

### **Ejercicio : 9**

**Una onda electromagnética plana se desplaza en el vacío en el sentido positivo del eje OX, con una frecuencia de 2·108 Hz. Calcular la longitud de onda y el periodo.**

---

In [ ]:
frecuencia = 2*108 #HZ

Cálculo de la longitud de onda

In [ ]:
longitud_de_onda = c / frecuencia

Cálculo del periodo

In [ ]:
periodo = 1 / frecuencia

In [ ]:
print(f'Longitud de onda: {longitud_de_onda} m')
print(f'Periodo: {periodo} s')

Longitud de onda: 1387928.0462962964 m
Periodo: 0.004629629629629629 s


### **Ejercicio : 10**
**Calcular la energía de los fotones correspondientes a los extremos, violeta y rojo, del espectro visible (400nm y 700nm)**

---

Convertimos las longitudes de ondas a metros

In [ ]:
lambda_violeta=400e-9
lambda_rojo=700e-9

Calculamos la energia $E = \frac{hc}{λ}$

In [ ]:
E_violeta = (h*c)/lambda_violeta
E_rojo = (h*c)/lambda_rojo

Resultado

In [ ]:
print(f'Energía del fotón violeta: {E_violeta} Joules')
print(f'Energía del fotón rojo: {E_rojo} Joules')

Energía del fotón violeta: 4.966114642872322e-19 Joules
Energía del fotón rojo: 2.837779795927041e-19 Joules


### **Ejercicio : 11**

**Dada una bola de billar de masa 1kg que se mueve con velocidad 0.25 m/s sobre la mesa de juego.
Según de Broglie ¿Cuál sería la longitud de onda asociada que llevaría esta bola? ¿Y si en lugar de
una bola de billar fuese un electrón?**

---

La relacion de Broglie estabelece que toda particular con un momento lineal (producto de su masa y velocidad) tiene asociada una longitud de onda dada por:
$ λ = \frac{h}{p}$
Donde p es el momento lineal de la particula, que se calcula como $p = m*v$ siendo (m) la masa y (v) la velocidad

Entonces para la bola de billar:

In [ ]:
m_bola = 1#kg
v = 0.25 #m/s
p_bola = m_bola*v
lambda_bola = h/p_bola
print(f"Longitud de onda para la bola de billar: {lambda_bola} m/s")

Longitud de onda para la bola de billar: 2.65042806e-33 m/s


In [ ]:
p_electron = electron_mass*v
lambda_electron = h/p_electron
print(f"Longitud de onda para la bola de electrón: {lambda_electron} m/s")

Longitud de onda para la bola de electrón: 0.0029095580413014835 m/s


### **Ejercicio : 12**

**La longitud de onda mínima para arrancar un electrón en una placa de potasio es de 564 nm.
Calcular la función del trabajo del Potasio. ¿Cuál es el potencial de detención cuando se ilumina la
placa con luz de 400nm?**

---

Para resolver esta pregunta, usaremos la ecuación del efecto fotoeléctrico:

$ h*f = ϕ + K_{max}$


*   h es la constante de Planck
*   f es la frecuencia de la luz incidente.
* ϕ es la función de trabajo del material, en este caso, potasio.

*   $K_{max}$  es la energía cinética máxima del electrón liberado.


La relación entre la frecuencia f y la longitud de onda λ es:
$$f = \frac{c}{λ}$$

Cuando se usa la longitud de onda mínima para arrancar un electrón, $K_{max}$ es 0.

Entonces nos queda la siguiente formula:
$$ ϕ = h*\frac{c}{λ}$$


In [ ]:
phi = (h *c)/564e-9#longitud de onda en metros
print(f"Función de trabajo del potasio: {phi} J")

Función de trabajo del potasio: 3.522067122604483e-19 J


Al iluminar la placa con luz de 400nm, determinamos el potencial de detencion con $K_{max} = e * V$ y la ecuacion del efecto fotoeléctrico se convierte en:
$$ h* f_{400nm} = ϕ + e * V $$ y esto en : $$V = \frac{h*f_{400nm}-ϕ}{e}$$

In [ ]:
v = (h*(c/400e-9)-phi)/(electron_mass)
print(f"Potencial de detención: {v} V")

Potencial de detención: 158523075499.6141 V


### **Ejercicio 13:                                                     Describir los siguientes experimentos y explicar cómo contribuyeron al desarrollo de la teoría cuántica**
---

#### **a)El Experimento de Stern-Gerlach**
---

**El Experimento de Stern-Gerlach**: Este experimento, realizado en 1922 por Otto Stern y Walther Gerlach, demostró que las propiedades cuánticas de las partículas, como el espín, tienen valores cuantizados. En el experimento, átomos de plata fueron enviados a través de un campo magnético no uniforme, lo que resultó en un haz dividido que correspondía a los átomos con espín hacia arriba y hacia abajo. Esta observación fue una de las primeras evidencias directas de la cuantización y una manifestación de que los sistemas cuánticos pueden tener solo ciertos valores discretos de ciertas cantidades físicas.


#### **b)El Efecto Compton**
---

**El Efecto Compton**: En 1923, Arthur Compton observó que los fotones dispersados por electrones tenían una longitud de onda más larga (menor energía) de lo esperado por la teoría clásica. Este efecto solo podía explicarse si la luz se comportaba como partículas (fotones) que colisionaban con los electrones, lo que significaba que la luz tenía propiedades tanto de onda como de partícula. El efecto Compton fue crucial para establecer la naturaleza dual onda-partícula de la luz y por extensión, la dualidad onda-partícula en la mecánica cuántica.



#### **c)El Efecto fotoeléctrico**
---

**El Efecto Fotoeléctrico**: Observado por primera vez por Heinrich Hertz en 1887 y explicado por Albert Einstein en 1905, el efecto fotoeléctrico es el fenómeno por el cual los electrones son emitidos desde una superficie de metal cuando esta es iluminada por luz de una frecuencia suficientemente alta. Einstein postuló que la luz consistía de cuantos de energía, o fotones, cuya energía estaba directamente relacionada con su frecuencia. Este concepto fue fundamental para la noción de que la energía en la escala cuántica viene en paquetes discretos (cuantos), una idea central en la teoría cuántica.

## **La Ecuación de Schrödinger**
---

### **Ejercicio 14: Dadas las funciones:**
\begin{align*}
a) \ \psi(x) &= x \\
b) \ \phi(x) &= e^{-x^2}
\end{align*}
\begin{align*}
\text{definidas en el intervalo } 0 \leq x \leq  \infty. \text{ Comprobar si pueden ser funciones de onda.}
\end{align*}

---

In [ ]:
# Función psi(x) = x
psi = lambda x: x**2
# Función phi(x) = e^(-x^2)
phi = lambda x: np.exp(-x**2)**2

# Calculamos las integrales de |psi(x)|^2 y |phi(x)|^2 en el intervalo [0, ∞)
integral_psi, error_psi = quad(psi, 0, np.inf)
integral_phi, error_phi = quad(phi, 0, np.inf)

(integral_psi, integral_phi), (error_psi, error_phi)



<ipython-input-34-499af40d0d48>:7: IntegrationWarning: The integral is probably divergent, or slowly convergent.
  integral_psi, error_psi = quad(psi, 0, np.inf)


((-0.33156386027260937, 0.6266570686577508),
 (0.0014214591807548138, 2.2337251559046266e-09))



Las integrales de las funciones dadas son:

- Para $( \psi(x) = x )$, la integral de $( |\psi(x)|^2 )$ en el intervalo $( [0, \infty) )$ no converge, lo que indica que la función no es normalizable y, por lo tanto, no puede representar una función de onda en mecánica cuántica.

- Para $( \phi(x) = e^{-x^2} )$, la integral de $( |\phi(x)|^2 )$ en el intervalo $( [0, \infty) )$ converge a aproximadamente 0.6267, lo que sugiere que esta función podría ser normalizable si se multiplica por el factor de normalización adecuado.

El resultado negativo para la integral de $( |\psi(x)|^2 )$ es probablemente debido a una advertencia de convergencia lenta o divergente emitida por la función de integración numérica. Esto es de esperarse ya que matemáticamente sabemos que la integral de $( x^2 )$ desde 0 a infinito es infinita.

### **Ejercicio 15:**

La función de onda para una partícula confinada en el intervalo **0 ≤ x ≤ a** en el estado fundamental es:

$$
\psi(x) = A \cdot \sin\left(\frac{\pi x}{a}\right)
$$

donde A es la constante de normalización. Encontrar:

---



#### **a) Encontrar el valor de A**
---

Definimos las variables que vamos a usar:

In [ ]:
A, x, a = symbols('A x a')

Definimos la función de onda:

In [ ]:
psi = A * sin(pi*x/a)

In [ ]:
integrand = (psi**2).integrate((x, 0, a))

In [ ]:
integrand

Piecewise((A**2*a/2, (a > -oo) & (a < oo) & Ne(a, 0)), (0, True))

In [ ]:
# Resolver para A
A_value = solve(integrand -1, A)[0]# -1 ya que el resultado tiene que ser = 1 y para resolver esta ecuacion haciendo eso la igualamos a 0
A_value

Piecewise((-sqrt(2)*sqrt(1/a), (a > -oo) & (a < oo) & Ne(a, 0)), (nan, True))

#### b) **Calcular la probabilidad de que la partícula se encuentre en el intervalo**
$$
 \frac{a}{2} \leq x \leq \frac{3a}{4}
$$

---

Sabiendo que  $A = -\sqrt{\frac{2}{a}} $

Entonces:

$|\psi(x)|^2 = \frac{2}{a} \sin^2\left(\frac{\pi x}{a}\right)$


Definir la función de onda con la constante de normalización ya encontrada

In [ ]:
psi_normalized = sqrt(2/a) * sin(pi*x/a)

Densidad de probabilidad

In [ ]:
probability_density = psi_normalized**2

In [ ]:
probability_density

2*sin(pi*x/a)**2/a

Integramos en el intervalo [a/2, 3a/4]

In [ ]:
probability = integrate(probability_density, (x, a/2, (3*a)/4))

In [ ]:
probability

Piecewise((-1/2 + 2*(1/4 + 3*pi/8)/pi, (a > -oo) & (a < oo) & Ne(a, 0)), (0, True))

### **Ejercicio 16: Dada una partícula cuya función de onda es de la forma: $\psi(x,t) = \phi(x) \chi(t) $ y sometida a un potencial $ V(x) :$**
---


#### **a) Demostrar que la parte temporal es del tipo  $\chi(t) = Ae^{i\omega t} $ siendo  A una constante**
---




Sustituyendo $(\psi(x,t) = \phi(x) \chi(t))$ en la ecuación de Schrödinger dependiente del tiempo, obtenemos:

$ i\hbar \phi(x) \frac{\partial \chi(t)}{\partial t} = -\frac{\hbar^2}{2m} \chi(t) \frac{\partial^2 \phi(x)}{\partial x^2} + V(x) \phi(x) \chi(t) $

Dividiendo ambos lados de la ecuación por $\phi(x)\chi(t)$, obtenemos:

$ i\hbar \frac{1}{\chi(t)} \frac{\partial \chi(t)}{\partial t} = -\frac{\hbar^2}{2m \phi(x)} \frac{\partial^2 \phi(x)}{\partial x^2} + V(x)$

El lado izquierdo de la ecuación es una función solo del tiempo, y el lado derecho es una función solo del espacio. Para que esta ecuación se mantenga verdadera para todos los valores de $(x) y $(t)$, ambos lados deben ser iguales a una constante. Llamemos a esta constante \(E\) (la energía de la partícula). Entonces, la ecuación para la parte temporal se convierte en:

$ i\hbar \frac{1}{\chi(t)} \frac{\partial \chi(t)}{\partial t} = E$

Resolviendo esta ecuación diferencial, encontramos que $(\chi(t) = Ae^{i\omega t})$, donde $(A)$ es una constante y $( \hbar \omega = E )$

#### **b) Demostrar que la parte espacial $ \phi(x) $ verifica la ecuación de Schrödinger independiente del tiempo con $ E = \hbar \omega $, es decir: $$\frac{-\hbar^2}{2m} \frac{d^2\phi(x)}{dx^2} + V(x)\phi(x) = \hbar\omega\phi(x) $$**
---



 **Demostración de la ecuación para $(\phi(x))$:**

De manera similar, la ecuación para la parte espacial se convierte en:

$ -\frac{\hbar^2}{2m} \frac{\partial^2 \phi(x)}{\partial x^2} + V(x)\phi(x) = E \phi(x) $

Esta es la ecuación de Schrödinger independiente del tiempo, con $(E = \hbar \omega)$.

#### **c) Demostrar que la densidad de probabilidad de esta función de onda es independiente del tiempo.**
---

La densidad de probabilidad está dada por $(|\psi(x,t)|^2)$. Sustituyendo $(\psi(x,t) = \phi(x) \chi(t))$, obtenemos:

$|\psi(x,t)|^2 = |\phi(x) \chi(t)|^2 = |\phi(x)|^2 |\chi(t)|^2 $

Ya que $(\chi(t) = Ae^{i\omega t})$, $(|\chi(t)|^2)$ es constante (ya que $(|Ae^{i\omega t}|^2 = |A|^2))$, lo que significa que la densidad de probabilidad $(|\psi(x,t)|^2)$ es independiente del tiempo.

### **Ejercicio 17:**

**Una partícula viene descrita por la siguiente función de onda:**
$$
\psi(x,t) = \sqrt{\frac{1}{a}} \sin\left(\frac{\pi x}{a}\right) e^{-iE_1t/\hbar} + \sqrt{\frac{1}{a}} \sin\left(\frac{2\pi x}{a}\right) e^{-iE_2t/\hbar}
$$

**Encontrar la corriente de probabilidad para este sistema.**

---

In [ ]:
# Definición de los símbolos necesarios
x, t, a, m, E1, E2 = sp.symbols('x t a m E1 E2', real=True, positive=True)
hbar_symbol = sp.symbols('hbar', real=True, positive=True)

# Definición de la función de onda ψ(x,t)
psi = sp.sqrt(1/a) * sp.sin(sp.pi * x / a) * sp.exp(-sp.I * E1 * t / hbar_symbol) + \
      sp.sqrt(1/a) * sp.sin(2 * sp.pi * x / a) * sp.exp(-sp.I * E2 * t / hbar_symbol)

# Derivada parcial de ψ(x,t) con respecto a x
psi_dx = sp.diff(psi, x)

# Complejo conjugado de ψ(x,t)
psi_star = sp.conjugate(psi)

# Producto de ψ*(x,t) y la derivada parcial de ψ(x,t) con respecto a x
product = psi_star * psi_dx

# Parte imaginaria del producto
imaginary_part = sp.im(product)

# Corriente de probabilidad J
J = (hbar_symbol / m) * imaginary_part

# Simplificamos la expresión
J = sp.simplify(J)

J


-2*pi*hbar*sin(pi*x/a)**3*sin(t*(E1 - E2)/hbar)/(a**2*m)

###**Ejercicio 18:**

Encontrar el valor $Δ x \Delta p$ para una partícula de masa \(m\) que se encuentra confinada en una caja de potencial en su estado fundamental. Su función de onda en estas condiciones es:
$$
\psi(x) = \sqrt{\frac{2}{a}} \sin\left(\frac{\pi x}{a}\right)
$$

---

In [ ]:

# Definiendo las variables
x, a = symbols('x a', real=True, positive=True)
psi = sqrt(2/a) * sin(pi*x/a)

# Calculando <x> y <x^2>
x_expect = integrate(psi * x * psi, (x, 0, a))
x2_expect = integrate(psi * x**2 * psi, (x, 0, a))

# Calculando la incertidumbre en x, Δx
delta_x = sqrt(x2_expect - x_expect**2)

# Calculando <p> y <p^2>
# Dado que psi es real, su conjugado complejo es psi, y <p> = 0
p_expect = 0
# El operador p^2 es -hbar^2 * d^2/dx^2
p2_expect = integrate(psi * (-hbar**2 * diff(psi, x, x)), (x, 0, a))

# Calculando la incertidumbre en p, Δp
delta_p = sqrt(p2_expect - p_expect**2)

# Producto Δx Δp
delta_x_delta_p = delta_x * delta_p

simplify(delta_x_delta_p), x_expect, x2_expect, delta_x, p_expect, p2_expect, delta_p


(1.75761969607693e-35*sqrt(-18 + 3*pi**2),
 a/2,
 2*(-a**3/(4*pi**2) + a**3/6)/a,
 sqrt(-a**2/4 + 2*(-a**3/(4*pi**2) + a**3/6)/a),
 0,
 1.11212171857352e-68*pi**2/a**2,
 1.05457181764616e-34*pi/a)

1. $ \Delta x \Delta p = 1.75761969607693 \times 10^{-35} \sqrt{-18 + 3\pi^2} $
2. $\langle x \rangle = \frac{a}{2} $
3. $\langle x^2 \rangle = \frac{2}{a}\left(-\frac{a^3}{4\pi^2} + \frac{a^3}{6}\right) $
4. $\Delta x = \sqrt{-\frac{a^2}{4} + \frac{2}{a}\left(-\frac{a^3}{4\pi^2} + \frac{a^3}{6}\right)} $
5. $ \langle p \rangle = 0 $
6. $\langle p^2 \rangle = \frac{1.11212171857352 \times 10^{-68} \pi^2}{a^2}$
7. $\Delta p = \frac{1.05457181764616 \times 10^{-34} \pi}{a} $


### **Ejercicio 19:**
La masa de una bala es de 50g y la de un electrón es de $(9,1 \times 10^{-28}g)$. Para ambos se mide una velocidad de \(300 m/s\) con una incertidumbre del \(0,01\%\).

Calcula la exactitud con la que puede medirse la posición de la bala y la del electrón si tanto velocidad como posición se miden de forma simultánea en el mismo experimento.

---


In [ ]:
# Constantes dadas
velocidad = 300  # m/s
incertidumbre_relativa = 0.01 / 100  # 0,01%
masa_bala = 50e-3  # kg (convertido de 50g a kg)
masa_electron = 9.1e-31  # kg (convertido de g a kg)

# Incertidumbre en la velocidad
delta_v = velocidad * incertidumbre_relativa

# Calculamos la incertidumbre en la posición usando el principio de incertidumbre de Heisenberg
# Δx Δp >= ħ/2 => Δx >= ħ / (2 * Δp) => Δx >= ħ / (2 * m * Δv)

# Constante de Planck reducida (en J*s)
hbar = const.hbar

# Incertidumbre en la posición de la bala y el electrón
delta_x_bala = hbar / (2 * masa_bala * delta_v)
delta_x_electron = hbar / (2 * masa_electron * delta_v)

delta_x_bala, delta_x_electron



(3.5152393921538546e-32, 0.0019314502154691507)


La incertidumbre en la posición para la bala es aproximadamente $( 3.52 \times 10^{-32} )$ metros y para el electrón es aproximadamente $( 1.93 \times 10^{-3} )$ metros o $1.93  milímetros$.

Estos resultados nos muestran la relación inversa entre la masa y la incertidumbre en la posición: debido a que la bala tiene una masa mucho mayor que el electrón, la incertidumbre en su posición es extremadamente pequeña, mientras que para el electrón, con una masa mucho menor, la incertidumbre en la posición es relativamente grande.

### **Ejercicio 20:**NIDEA
Calcular los valores esperados $(<x>)$ y $(<p>)$ para una partícula confinada en una caja unidimensional de longitud $(0 \leq x \leq a)$ que se encuentra en su estado fundamental descrito por la función de onda:
$$
\psi(x) = \sqrt{\frac{2}{a}} \sin\left(\frac{\pi x}{a}\right)
$$

---

In [64]:
# Definir las variables simbólicas
x, a, h_bar = symbols('x a h_bar', real=True, positive=True)

# Definir la función de onda
psi_x = sp.sqrt(2/a) * sp.sin(pi*x/a)

# Calcular el valor esperado de x
valor_esperado_x = integrate(x * psi_x**2, (x, 0, a))

# Calcular el valor esperado de p
derivada_psi_x = sp.diff(psi_x, x)
valor_esperado_p = integrate(psi_x * (-I * h_bar * derivada_psi_x), (x, 0, a))

valor_esperado_x, valor_esperado_p.simplify()


(a/2, 0)

Valor esperado de la posición $(\langle x \rangle)$:

$$
\langle x \rangle = \frac{a}{2}
$$

Esto es lo que esperaríamos, ya que la partícula tiene la misma probabilidad de estar en cualquier lugar de la caja, por lo que el valor promedio de su posición es justo en el medio de la caja.

Valor esperado del momento $(\langle p \rangle)$:

$$
\langle p \rangle = 0
$$

Esto se debe a que la función de onda es simétrica respecto al centro de la caja, lo que implica que cualquier impulso neto en una dirección está equilibrado por un impulso igual y opuesto en la otra dirección, resultando en un valor esperado de momento cero.


### **Ejercicio 21:**
Hallar el valor esperado de la posición de un oscilador armónico cuántico en el estado fundamental.

---

Para hallar el valor esperado de la posición $\langle x \rangle$ de un oscilador armónico cuántico en el estado fundamental, se utiliza la definición del valor esperado en mecánica cuántica:

$$ \langle x \rangle = \langle 0 | \hat{x} | 0 \rangle $$

donde $|0\rangle$ es el estado fundamental del oscilador armónico y $\hat{x}$ es el operador de posición. En el estado fundamental, la función de onda es simétrica con respecto al origen, lo que lleva a que el valor esperado de la posición sea cero:

$$ \langle x \rangle = 0 $$

Esto se debe a la simetría de la función de onda del estado fundamental y a la antisimetría del operador de posición. La función de onda del estado fundamental del oscilador armónico cuántico es:

$$ \psi_0(x) = \left(\frac{m\omega}{\pi\hbar}\right)^{1/4} e^{-\frac{m\omega x^2}{2\hbar}} $$

El valor esperado se calcula como la integral de la función de onda multiplicada por el operador de posición y la función de onda conjugada:

$$\langle x \rangle = \int_{-\infty}^{\infty} \psi_0^*(x) \, x \, \psi_0(x) \, dx $$

Dado que $\psi_0(x)$ es una función gaussiana y $x$ es un operador impar, la integral de una función impar sobre un intervalo simétrico es cero, lo que confirma que el valor esperado de la posición en el estado fundamental es cero.


# Mecánica

---

### **Ejercicio 1:**
Escribir las ecuaciones de Euler-Lagrange para una partícula que se mueve en el espacio bidimensional sometida a un campo de fuerzas $ \mathbf{F} $. Demostrar que dichas ecuaciones son equivalentes a la segunda ley de Newton.

---

1. **Lagrangiano de la Partícula**: El Lagrangiano $ L $ para una partícula en un campo de fuerzas es la diferencia entre su energía cinética $ T $ y su energía potencial $ V $:
   $$
   L = T - V
   $$

2. **Forma de la Energía Cinética y Potencial**: En un espacio bidimensional, la energía cinética $ T $ para una partícula de masa $ m $ y la energía potencial $ V $ son dadas por:
   $$
   T = \frac{1}{2} m (\dot{x}^2 + \dot{y}^2), \quad V = V(x, y)
   $$

3. **Ecuaciones de Euler-Lagrange**: Las ecuaciones de Euler-Lagrange son:
   $$
   \frac{d}{dt}\left(\frac{\partial L}{\partial \dot{q}_i}\right) - \frac{\partial L}{\partial q_i} = 0, \quad i = x, y
   $$

4. **Aplicación a las Coordenadas Bidimensionales**: Aplicando las ecuaciones de Euler-Lagrange a las coordenadas $ x $ y $ y $, obtenemos:
   $$
   \frac{d}{dt}\left(\frac{\partial L}{\partial \dot{x}}\right) - \frac{\partial L}{\partial x} = 0, \quad \frac{d}{dt}\left(\frac{\partial L}{\partial \dot{y}}\right) - \frac{\partial L}{\partial y} = 0
   $$

5. **Equivalencia con la Segunda Ley de Newton**: Al desarrollar estas ecuaciones y asumiendo que la fuerza $ F $ deriva de la energía potencial $ V $, se puede demostrar que estas ecuaciones son equivalentes a la segunda ley de Newton:
   $$
   m\ddot{x} = F_x, \quad m\ddot{y} = F_y
   $$

Las ecuaciones de Euler-Lagrange para una partícula en un campo de fuerzas bidimensional se reducen a la segunda ley de Newton cuando se consideran las fuerzas derivadas de un potencial.


### **Ejercicio 2:**
Demostrar que $ H=T+V $ (considerar las coordenadas espaciales de un sistema que se mueve en una dimensión).

---

1. **Definición de la Hamiltoniana**: La Hamiltoniana $ H $ de un sistema físico es la suma de su energía cinética $ T $ y su energía potencial $ V $:
   $$
   H = T + V
   $$

2. **Energía Cinética en Una Dimensión**: Para un sistema que se mueve en una dimensión, la energía cinética $ T $, para una partícula de masa $ m $ y velocidad $ v $, es dada por:
   $$
   T = \frac{1}{2} m v^2
   $$

3. **Energía Potencial**: La energía potencial $ V $ del sistema depende de la posición $ x $ en el campo de fuerzas y se expresa generalmente como $ V(x) $.

4. **Hamiltoniana en Una Dimensión**: Combinando las expresiones de $ T $ y $ V $, la Hamiltoniana en una dimensión se convierte en:
   $$
   H = \frac{1}{2} m v^2 + V(x)
   $$

5. **Conclusión**: Esta expresión muestra que la Hamiltoniana $ H $ es efectivamente la suma de las energías cinética y potencial del sistema en una dimensión.


### **Ejercicio 3:**
Escribir las ecuaciones de Hamilton para una partícula que se mueve en el espacio bidimensional sometida a un campo de fuerzas $ \mathbf{F} $. Demostrar que dichas ecuaciones son equivalentes a la segunda ley de Newton.

---

1. **Definición de la Hamiltoniana**: En un espacio bidimensional, la Hamiltoniana $( H )$ para una partícula bajo un campo de fuerzas $( \mathbf{F} )$ es la suma de la energía cinética $( T )$ y la energía potencial $( V )$. Si $( q_1, q_2 )$ son las coordenadas y $( p_1, p_2 )$ los momentos conjugados, la Hamiltoniana se expresa como:
   $$
   H = \frac{1}{2m}(p_1^2 + p_2^2) + V(q_1, q_2)
   $$

2. **Ecuaciones de Hamilton**: Las ecuaciones de Hamilton para este sistema son:
   $$
   \dot{q}_i = \frac{\partial H}{\partial p_i}, \quad \dot{p}_i = -\frac{\partial H}{\partial q_i}, \quad i = 1, 2
   $$
   Estas ecuaciones describen cómo cambian las coordenadas y los momentos con el tiempo.

3. **Aplicación a la Hamiltoniana**: Al aplicar estas ecuaciones a la Hamiltoniana definida, obtenemos:
   $$
   \dot{q}_1 = \frac{p_1}{m}, \quad \dot{q}_2 = \frac{p_2}{m}
   $$
   $$
   \dot{p}_1 = -\frac{\partial V}{\partial q_1}, \quad \dot{p}_2 = -\frac{\partial V}{\partial q_2}
   $$

4. **Equivalencia con la Segunda Ley de Newton**: La relación entre la fuerza y el potencial es $( \mathbf{F} = -\nabla V )$. Por lo tanto, las ecuaciones para $( \dot{p}_i )$ se pueden reescribir como $( m\ddot{q}_i = F_i )$, que es la forma de la segunda ley de Newton.

Las ecuaciones de Hamilton en este contexto describen cómo las coordenadas y los momentos de una partícula cambian en el tiempo y son equivalentes a la segunda ley de Newton al considerar la relación entre fuerza y potencial.

### **Ejercicio 4:**
Comprobar que el operador traslación espacial es unitario.

---

1. **Definición del Operador de Traslación**: Un operador de traslación espacial $ T(\mathbf{a}) $ desplaza un punto en el espacio por un vector $ \mathbf{a} $. En mecánica cuántica, este operador actúa sobre una función de onda $ \psi(\mathbf{r}) $ de la siguiente manera:
   $$
   T(\mathbf{a})\psi(\mathbf{r}) = \psi(\mathbf{r} - \mathbf{a})
   $$

2. **Definición de un Operador Unitario**: Un operador $ U $ es unitario si cumple con la condición $ U^\dagger U = U U^\dagger = I $, donde $ U^\dagger $ es el adjunto hermítico de $ U $ y $ I $ es el operador identidad.

3. **Calculando el Adjunto Hermítico**: El adjunto hermítico de $ T(\mathbf{a}) $, denotado como $ T^\dagger(\mathbf{a}) $, actúa sobre una función de onda $ \phi(\mathbf{r}) $ de la forma:
   $$
   T^\dagger(\mathbf{a})\phi(\mathbf{r}) = \phi(\mathbf{r} + \mathbf{a})
   $$

4. **Verificación de la Unitariedad**: Para verificar que $ T(\mathbf{a}) $ es unitario, debemos mostrar que $ T^\dagger(\mathbf{a}) T(\mathbf{a}) = T(\mathbf{a}) T^\dagger(\mathbf{a}) = I $. Esto significa que si aplicamos primero $ T(\mathbf{a}) $ y luego $ T^\dagger(\mathbf{a}) $ a cualquier función de onda $ \psi(\mathbf{r}) $, el resultado debe ser la función de onda original $ \psi(\mathbf{r}) $. Matemáticamente:
   $$
   T^\dagger(\mathbf{a})T(\mathbf{a})\psi(\mathbf{r}) = T^\dagger(\mathbf{a})\psi(\mathbf{r} - \mathbf{a}) = \psi(\mathbf{r})
   $$
   y de manera similar para $ T(\mathbf{a})T^\dagger(\mathbf{a}) $.

5. **Conclusión**: Dado que la composición de $ T(\mathbf{a}) $ y $ T^\dagger(\mathbf{a}) $ en cualquier orden devuelve la función de onda original, concluimos que el operador de traslación espacial $ T(\mathbf{a}) $ es unitario.


### **Ejercicio 6:**
Demostrar que cuando dos operadores conmutan tienen una base propia común.

---

Supongamos dos operadores hermíticos $ A $ y $ B $ que conmutan, es decir, $ AB = BA $. Queremos demostrar que tienen una base de vectores propios común.

1. **Teorema Espectral para Operadores Hermíticos**: Dado que $ A $ y $ B $ son hermíticos, por el teorema espectral, cada uno tiene una base completa de vectores propios.

2. **Vectores Propios de A**: Sea $ |a\rangle $ un vector propio de $ A $, entonces $ A|a\rangle = a|a\rangle $, donde $ a $ es el valor propio correspondiente.

3. **Aplicación de B a $ |a\rangle $**: Aplicamos $ B $ a ambos lados de la ecuación anterior:
   $$
   B(A|a\rangle) = B(a|a\rangle) = a(B|a\rangle)
   $$

4. **Conmutatividad de A y B**: Usando la conmutatividad $ AB = BA $, también tenemos:
   $$
   A(B|a\rangle) = B(A|a\rangle) = a(B|a\rangle)
   $$

5. **$ B|a\rangle $ es Vector Propio de A**: De la ecuación anterior, se deduce que $ B|a\rangle $ es también un vector propio de $ A $ con el mismo valor propio $ a $.

6. **Base Común de Vectores Propios**: Repitiendo el argumento para todos los vectores propios de $ A $ y usando un razonamiento similar para $ B $, se muestra que los operadores comparten una base común de vectores propios.

La conmutatividad de dos operadores hermíticos implica que tienen una base de vectores propios común, lo cual es una consecuencia directa de su acción en los vectores propios del otro operador.


### **Ejercicio 7:**
Hallar los autovalores y autovectores del operador de espín $ S_x $

---

1. **Expresión Matricial de $ S_x $**: El operador de espín $ S_x $, en la base de los estados de espín z, se representa por la matriz:
   $$
   S_x = \frac{\hbar}{2}\begin{pmatrix} 0 & 1 \\ 1 & 0 \end{pmatrix}
   $$

2. **Ecuación Característica**: Para encontrar los autovalores, resolvemos la ecuación característica:
   $$
   \det(S_x - \lambda I) = 0
   $$

3. **Determinante y Autovalores**: El determinante se convierte en:
   $$
   \det\left(\frac{\hbar}{2}\begin{pmatrix} 0 & 1 \\ 1 & 0 \end{pmatrix} - \lambda \begin{pmatrix} 1 & 0 \\ 0 & 1 \end{pmatrix}\right) = 0
   $$
   Resolviendo esta ecuación cuadrática, obtenemos los autovalores $ \lambda = \pm\frac{\hbar}{2} $.

4. **Encontrar Autovectores**: Para cada autovalor $ \lambda $, resolvemos la ecuación:
   $$
   (S_x - \lambda I)|v\rangle = 0
   $$
   para encontrar los autovectores correspondientes.

5. **Solución para Autovectores**: La solución de estas ecuaciones nos dará los autovectores de $ S_x $, que son combinaciones lineales de los estados de espín en la dirección z.



In [61]:
# Definir la constante simbólica para ħ (h-barra)
h_bar = sp.symbols('h_bar')

# Matriz S_x
S_x = (h_bar/2) * sp.Matrix([[0, 1], [1, 0]])

# Autovalores
autovalores = S_x.eigenvals()

# Cálculo de los autovectores
autovectores = S_x.eigenvects()

autovalores, autovectores

({-h_bar/2: 1, h_bar/2: 1},
 [(-h_bar/2, 1, [Matrix([
    [-1],
    [ 1]])]),
  (h_bar/2,
   1,
   [Matrix([
    [1],
    [1]])])])

- Para $ \lambda = -\frac{\hbar}{2}$ , el autovector es $\begin{pmatrix} -1 \\ 1 \end{pmatrix}$
- Para $ \lambda = \frac{\hbar}{2} $, el autovector es $\begin{pmatrix} 1 \\ 1 \end{pmatrix}$

### **Ejercicio 8:**
Comprobar que los estados de Bell son máximamente entrelazados.

---

Los cuatro estados de Bell son:
1. $( |\Phi^+\rangle = \frac{1}{\sqrt{2}}(|00\rangle + |11\rangle) )$
2. $( |\Phi^-\rangle = \frac{1}{\sqrt{2}}(|00\rangle - |11\rangle) )$
3. $( |\Psi^+\rangle = \frac{1}{\sqrt{2}}(|01\rangle + |10\rangle) )$
4. $( |\Psi^-\rangle = \frac{1}{\sqrt{2}}(|01\rangle - |10\rangle) )$

Un estado entrelazado es "máximamente entrelazado" si no es posible escribirlo como un producto de estados de sus subsistemas. Un estado de dos qubits está máximamente entrelazado si su entropía de entrelazamiento es máxima, lo que significa que la información sobre un qubit es completamente dependiente del otro.

La entropía de entrelazamiento de un estado puro de dos qubits, $( |\psi\rangle )$, se define como la entropía de Von Neumann de cualquiera de los subsistemas. Para un estado de Bell, la entropía de Von Neumann es máxima. En el caso de los estados de Bell, la densidad reducida de uno de los qubits es $( \frac{1}{2}I )$, donde $( I )$ es la matriz identidad de 2x2. La entropía de Von Neumann de una matriz densidad $( \rho )$ se calcula como $( S(\rho) = -\text{tr}(\rho \log_2 \rho) )$. Para $( \frac{1}{2}I )$, esto se simplifica a 1 bit, que es el valor máximo posible para un qubit, confirmando que los estados de Bell son máximamente entrelazados.

### **Ejercicio 9:**
Comprobar que $ \frac{1}{\sqrt{2}}(|00\rangle + |11\rangle) = \frac{1}{\sqrt{2}}(|++\rangle + |--\rangle) $.

---

In [ ]:
# Definición de los estados base |0>, |1>, |+>, y |->
estado_0 = np.array([1, 0])
estado_1 = np.array([0, 1])

estado_mas = (1/np.sqrt(2)) * (estado_0 + estado_1)
estado_menos = (1/np.sqrt(2)) * (estado_0 - estado_1)

# Estado |00> + |11>
estado_izquierdo = (1/np.sqrt(2)) * (np.kron(estado_0, estado_0) + np.kron(estado_1, estado_1))

# Estado |++> + |-->
estado_derecho = (1/np.sqrt(2)) * (np.kron(estado_mas, estado_mas) + np.kron(estado_menos, estado_menos))

# Comprobación de igualdad
np.allclose(estado_izquierdo, estado_derecho)


True

### **Ejercicio 10:**
Dar la expresión del estado de Bell $ \frac{1}{\sqrt{2}}(|00\rangle - |11\rangle) $ después de que $ A $ y $ B $ hayan rotado sus respectivos qubits un ángulo genérico $ \phi $

---

In [ ]:
# Definición de un símbolo genérico para phi
phi = sp.Symbol('phi')

# Redefinición de la matriz de rotación
def matriz_rotacion_sympy(phi):
    return sp.Matrix([[sp.exp(-sp.I * phi / 2), 0],
                      [0, sp.exp(sp.I * phi / 2)]])

# Estados base |0> y |1>
estado_0_sympy = sp.Matrix([1, 0])
estado_1_sympy = sp.Matrix([0, 1])

# Estado de Bell
estado_bell_sympy = (1/sp.sqrt(2)) * (sp.Matrix([1, 0, 0, 0]) - sp.Matrix([0, 0, 0, 1]))

# Aplicación de la matriz de rotación a ambos qubits
R_phi_sympy = matriz_rotacion_sympy(phi)
estado_transformado_sympy = sp.kronecker_product(R_phi_sympy, R_phi_sympy) * estado_bell_sympy

estado_transformado_sympy

Matrix([
[sqrt(2)*exp(-I*phi)/2],
[                    0],
[                    0],
[-sqrt(2)*exp(I*phi)/2]])